endpoints/openrtb2/amp_auction.go

```diff
@@ -0,0 +1,230 @@
+	} else if usersyncs.LiveSyncCount() == 0 {
+		deps.metrics.NoCookieMeter.Mark(1)
+		if isSafari {
+			deps.metrics.SafariNoCookieMeter.Mark(1)
+		}
+	}
+
+	response, err := deps.ex.HoldAuction(ctx, req, usersyncs)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "Critical error while running the auction: %v", err)
+		glog.Errorf("/openrtb2/amp Critical error: %v", err)
+		return
+	}
+
+	// Need to extract the targeting parameters from the response, as those are all that
+	// go in the AMP response
+	targets := map[string]string{}
+	for _, seatBids := range response.SeatBid {
+		for _, bid := range seatBids.Bid {
+			if strings.Contains(string(bid.Ext), "\"hb_cache_id") {
+				// Looking for cache_id to be set, as this should only be set on winning bids (or
+				// deal bids), and AMP can only deliver cached ads in any case.
+				// Note, this could casue issues if a targeting key value starts with "hb_cache_id",
+				// but this is a very unlikely corner case. Doing this so we can catch "hb_cache_id"
+				// and "hb_cache_id_{deal}", which allows for deal support in AMP.
+				bidExt := &openrtb_ext.ExtBid{}
+				err := json.Unmarshal(bid.Ext, bidExt)
+				if err != nil {
+					w.WriteHeader(http.StatusInternalServerError)
+					fmt.Fprintf(w, "Critical error while unpacking AMP targets: %v", err)
+					glog.Errorf("/openrtb2/amp Critical error unpacking targets: %v", err)
+					return
+				}
+				for key, value := range bidExt.Prebid.Targeting {
+					targets[key] = value
+				}
+			}
+		}
+	}
+
+	// Now JSONify the tragets for the AMP response.
+	ampResponse := AmpResponse{
+		Targeting: targets,
+	}
+
+	// Fixes #231
+	enc := json.NewEncoder(w)
+	enc.SetEscapeHTML(false)
+
+	// If an error happens when encoding the response, there isn't much we can do.
+	// If we've sent _any_ bytes, then Go would have sent the 200 status code first.
+	// That status code can't be un-sent... so the best we can do is log the error.
+	if err := enc.Encode(ampResponse); err != nil {
+		glog.Errorf("/openrtb2/amp Error encoding response: %v", err)
+	}
+}
+
+// parseRequest turns the HTTP request into an OpenRTB request.
+// If the errors list is empty, then the returned request will be valid according to the OpenRTB 2.5 spec.
+// In case of "strong recommendations" in the spec, it tends to be restrictive. If a better workaround is
+// possible, it will return errors with messages that suggest improvements.
+//
+// If the errors list has at least one element, then no guarantees are made about the returned request.
+func (deps *endpointDeps) parseAmpRequest(httpRequest *http.Request) (req *openrtb.BidRequest, errs []error) {
+	req = &openrtb.BidRequest{}
+	errs = nil
+
+	ampId := httpRequest.FormValue("tag_id")
+	if len(ampId) == 0 {
+		errs = []error{errors.New("AMP requests require an AMP config ID")}
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(50)*time.Millisecond)
```

Comment:
As a last minute change to my PR, I made this "50" into a constant... should probably use it here too.